flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/metadata/MetadataV5Serializer.java

```diff
@@ -0,0 +1,109 @@
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.checkpoint.metadata;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.runtime.checkpoint.FinishedOperatorSubtaskState;
+import org.apache.flink.runtime.checkpoint.FullyFinishedOperatorState;
+import org.apache.flink.runtime.checkpoint.OperatorState;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.state.memory.ByteStreamStateHandle;
+
+import javax.annotation.Nullable;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+
+import static org.apache.flink.util.Preconditions.checkState;
+
+/** V5 serializer that adds Operator name and uid. */
+@Internal
+public class MetadataV5Serializer extends MetadataV4Serializer {
+
+    public static final MetadataSerializer INSTANCE = new MetadataV5Serializer();
+    public static final int VERSION = 5;
+
+    @Override
+    public int getVersion() {
+        return VERSION;
+    }
+
+    @Override
+    protected void serializeOperatorState(OperatorState operatorState, DataOutputStream dos)
+            throws IOException {
+        // Name and UID are null in the whole chain when not provided, but since we can store
+        // strings in metadata we do conversion here
+        dos.writeUTF(operatorState.getOperatorName().orElse(""));
+        dos.writeUTF(operatorState.getOperatorUid().orElse(""));
+        super.serializeOperatorState(operatorState, dos);
+    }
+
+    @Override
+    protected OperatorState deserializeOperatorState(
+            DataInputStream dis, @Nullable DeserializationContext context) throws IOException {
+        final String name = dis.readUTF();
+        final String uid = dis.readUTF();
+
+        final OperatorID jobVertexId = new OperatorID(dis.readLong(), dis.readLong());
+        final int parallelism = dis.readInt();
+        final int maxParallelism = dis.readInt();
+
+        ByteStreamStateHandle coordinateState =
+                deserializeAndCheckByteStreamStateHandle(dis, context);
+
+        final int numSubTaskStates = dis.readInt();
+        if (numSubTaskStates < 0) {
+            checkState(
+                    coordinateState == null,
+                    "Coordinator State should be null for fully finished operator state");
+            return new FullyFinishedOperatorState(
+                    name, uid, jobVertexId, parallelism, maxParallelism);
```

Comment:
The deserialized `name` and `uid` could be empty string right? Is it better be changed to `null` if so?

Reply:
Fair point, couple of lines later there is such a construct with needs to be applied here:
```
                        name.isEmpty() ? null : name,
                        uid.isEmpty() ? null : uid,
```

Reply:
I thought about it again. Shall we make the (de)serialization of `name` and `uid` null-aware? I'm worried that the user may specify an empty string for the uid or name, and after deserialization it turns to `null`. WDYT?

Reply:
I'm not totally against to add some sort of nullable string but I've the feeling that we would take unnecessary risk + more complexity for zero user value. I personally think that empty string is not a valid value (maybe we can just block that). The whole reason why we have such field is to help users to name operators in a human readable name.

Reply:
Yes, blocking empty string is a good option. Will you add those in this PR?

And is it possible add a UT for `MetadataV5Serializer`?

Reply:
Sure, will add soon.

Reply:
Added + rebased to the latest master.